### ISSUE TITLE
`unnecessary_unwrap`  false negative of Option variable in struct

### ISSUE THREAD

**Issue by gwen-lg** [rust-lang/rust-clippy#15321]
### Summary

If an Option variable tested with if + is_none before unwrap is local, the lint `unnecessary_unwrap` is reported. But if the variable is nested in a struct, nothing is reported.

### Lint Name

unnecessary_unwrap

### Reproducer

I tried this code:

```rust
    struct Soption {
        option: Option<bool>,
    }
    let sopt = Soption { option: Some(true) };
    let _res = if sopt.option.is_none() {
        false // default
    } else {
        sopt.option.unwrap()
    };
```

I expected to see this happen:
```
warning: called `unwrap` on `option` after checking its variant with `is_none`
 --> src/clippy_unwrap_or.rs:7:9
  |
4 |     let _res = if sopt.option.is_none() {
  |                ------------------- help: try: `if let Some(<item>) = option`
...
7 |         sopt.option.unwrap()
  |         ^^^^^^^^^^^^^^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_unwrap
```

Instead, this happened:
None

while if the option var is local, a report is provided :
```rust
    let opt = Some(true);
    let _res = if opt.is_none() {
        false // default
    } else {
        opt.unwrap()
    };
```
```
warning: called `unwrap` on `opt` after checking its variant with `is_none`
  --> src/clippy_unwrap_or.rs:14:9
   |
11 |     let _res = if opt.is_none() {
   |                ---------------- help: try: `if let Some(<item>) = opt`
...
14 |         opt.unwrap()
   |         ^^^^^^^^^^^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_unwrap
   = note: `#[warn(clippy::unnecessary_unwrap)]` on by default
```

### Version

```text
rustc 1.88.0 (6b00bc388 2025-06-23)
binary: rustc
commit-hash: 6b00bc3880198600130e1cf62b8f8a93494488cc
commit-date: 2025-06-23
host: x86_64-unknown-linux-gnu
release: 1.88.0
LLVM version: 20.1.5
```

**Comment by ada4a** [comment]
I got confused by the last codeblock, because it shows the lint message from `unnecessary_literal_unwrap`, not `unnecessary_unwrap`. Here's the correct message:
```
warning: called `unwrap` on `opt` after checking its variant with `is_none`
 --> src/lib.rs:7:9
   |
7  |     let _res = if opt.is_none() {
   |                ---------------- help: try: `if let Some(<item>) = opt`
...
10 |         opt.unwrap()
   |         ^^^^^^^^^^^^
   |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_unwrap
```

**Comment by gwen-lg** [comment]
@ada4a : Yes, I made a mistake copying, sorry.
It's now fixed.
